ledger-service/http-json/src/test/scala/com/digitalasset/http/ContractsFetchTest.scala

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.http
+
+import com.digitalasset.daml.lf.data.FlatSpecCheckLaws
+import ContractsFetch.InsertDeleteStep
+
+import org.scalatest.prop.GeneratorDrivenPropertyChecks
+import org.scalatest.{FlatSpec, Matchers}
+import scalaz.{Equal, Monoid}
+import scalaz.syntax.semigroup._
+import scalaz.scalacheck.ScalazProperties
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+class ContractsFetchTest
+    extends FlatSpec
+    with Matchers
+    with FlatSpecCheckLaws
+    with GeneratorDrivenPropertyChecks {
+  import ContractsFetchTest._
+
+  behavior of "InsertDeleteStep append monoid"
+
+  checkLaws(ScalazProperties.monoid.laws[IDS])
+
+  behavior of "InsertDeleteStep.appendWithCid"
+
+  it should "never insert a deleted item" in forAll { (x: IDS, y: IDS) =>
+    val xy = x |+| y.copy(inserts = y.inserts filterNot x.deletes)
+    xy.inserts.toSet intersect xy.deletes shouldBe empty
+  }
+
+  it should "preserve every left delete" in forAll { (x: IDS, y: IDS) =>
+    val xy = x |+| y
+    // preservation of right deletes is an accident, not an invariant
```

Comment:
why is it an accident? we want all right deletes in the `xy`, don't we? these can be deletes for the inserts from the previous batch.

Reply:
If a right delete is in the left inserts, whether it is present or not is semantically irrelevant. I added a test in ac4b035293 on account of this.